rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+    }
+}
+
+impl Drop for PgsqlTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PgsqlStateProgress {
+    IdleState,
+    SSLRequestReceived,
+    SSLAcceptedReceived,
+    SSLRejectedReceived,
+    StartupMessageReceived,
+    SASLAuthenticationReceived,
+    SASLInitialResponseReceived,
+    // SSPIAuthenticationReceived, // TODO implement
+    SASLResponseReceived,
+    SASLAuthenticationFinalReceived,
+    SimpleAuthenticationReceived,
+    PasswordMessageReceived,
+    AuthenticationOkReceived,
+    ParameterSetup,
+    ReadyForQueryReceived,
+    SimpleQueryReceived,
+    RowDescriptionReceived,
+    DataRowReceived,
+    CommandCompletedReceived,
+    ErrorMessageReceived, // TODO do we really need this one?
+    ConnectionTerminated,
+    UnknownState, // TODO do we really need this one?
+    Finished,
+}
+
+pub struct PgsqlState {
+    tx_id: u64,
+    transactions: Vec<PgsqlTransaction>,
+    request_gap: bool,
+    response_gap: bool,
+    backend_secret_key: u32,
+    backend_pid: u32,
+    state_progress: PgsqlStateProgress,
+}
+
+impl PgsqlState {
+    pub fn new() -> Self {
+        Self {
+            tx_id: 0,
+            transactions: Vec::new(),
+            request_gap: false,
+            response_gap: false,
+            backend_secret_key: 0,
+            backend_pid: 0,
+            state_progress: PgsqlStateProgress::IdleState,
+        }
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&PgsqlTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self) -> PgsqlTransaction {
+        let mut tx = PgsqlTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        SCLogDebug!("Creating new transaction. tx_id: {}", tx.tx_id);
+        return tx;
+    }
+
+    /// Find or create a new transaction
+    ///
+    /// If a new transaction is created, push that into state.transactions before returning &mut to last tx
+    /// If we can't find a transaction and we should not create one, we return None
+    /// The moment when this is called will may impact the logic of transaction tracking (e.g. when a tx is considered completed)
+    // TODO A future, improved version may be based on current message type and dir, too
+    fn find_or_create_tx(&mut self) -> Option<&mut PgsqlTransaction> {
```

Comment:
We're not, entirely, right. I followed the default, somehow. What we do is... certain StateProgresses assumedly should have a transaction, because of how Pgsql works - and basically when we are parsing a response, to be more precise. In such case, and if this isn't a point where we should create a new transaction, the last one is the one we want. That's the one we "find".